test/topology_random_failures/test_random_failures.py

```diff
@@ -0,0 +1,212 @@
+async def test_random_failures(manager: ManagerClient,
+                               random_tables: RandomTables,
+                               error_injection: str,
+                               cluster_event: ClusterEventType) -> None:
+    table = await random_tables.add_table(ncolumns=5)
+    await table.insert_seq()
+
+    cluster_event_steps = cluster_event(manager, random_tables, error_injection)
+
+    LOGGER.info("Run preparation step of the cluster event.")
+    await anext(cluster_event_steps)
+
+    if error_injection == "stop_after_updating_cdc_generation":
+        # This error injection is a special one and should be handled on a coordinator node:
+        #   1. Enable the injection on a coordinator node
+        #   2. Bootstrap a new node
+        #   3. Wait till the injection handler will print the message to the log
+        #   4. Pause the bootstrapping node
+        #   5. Send the message to the injection handler to continue
+        coordinator = await get_coordinator_host(manager=manager)
+        await manager.api.enable_injection(node_ip=coordinator.ip_addr, injection=error_injection, one_shot=True)
+        coordinator_log = await manager.server_open_log(server_id=coordinator.server_id)
+        coordinator_log_mark = await coordinator_log.mark()
+        s_info = await manager.server_add(expected_server_state=ServerState.PROCESS_STARTED)
+        await coordinator_log.wait_for(
+            pattern="stop_after_updating_cdc_generation: wait for message for 5 minutes",
+            from_mark=coordinator_log_mark,
+        )
+        await manager.server_pause(server_id=s_info.server_id)
+        await manager.api.message_injection(node_ip=coordinator.ip_addr, injection=error_injection)
+    else:
+        s_info = await manager.server_add(
+            config={"error_injections_at_startup": [{"name": error_injection, "one_shot": True}]},
+            expected_server_state=ServerState.PROCESS_STARTED,
+        )
+
+    LOGGER.info("Wait till the bootstrapping node will be paused.")
+    await manager.wait_for_scylla_process_status(
+        server_id=s_info.server_id,
+        expected_statuses=[
+            psutil.STATUS_STOPPED,
+        ],
+    )
+
+    LOGGER.info("Run the cluster event main step.")
+    try:
+        cluster_event_start = time.perf_counter()
+        await anext(cluster_event_steps)
+        cluster_event_duration = time.perf_counter() - cluster_event_start
+        LOGGER.info("Cluster event `%s' took %.1fs", cluster_event.__name__, cluster_event_duration)
+    finally:
+        LOGGER.info("Unpause the server and wait till it wake up and become connectable using CQL.")
+        await manager.server_unpause(server_id=s_info.server_id)
+
+    server_log = await manager.server_open_log(server_id=s_info.server_id)
+
+    if cluster_event_duration + 1 >= WAIT_FOR_IP_TIMEOUT and error_injection in (  # give one more second for a tolerance
```

Comment:
I smell flakiness

Reply:
Do you have any ideas how to make it more bulletproof?

Reply:
Can you explain the reasoning behind this logic first? Then I'll try to come up with something.

Reply:
This code handle a situation when the added node sent a join request and then do not respond for 30+ second (which is wait_for_ip timeout) In this case the new node just hang and we need shut down it manually.

So, I just measure the time spent on a cluster event and if it close to 30 seconds I check coordinator's logs for the message about the node.

Reply:
Hm, ok I think I understand.

If the cluster event took 30+ seconds for whatever reason, then there is a **high chance** (but not a guarantee) that the topology coordinator timed out inside "wait for IP" and this will cause the booting node to hang forever after we unpause it.

But it could be due to process scheduling or whatever, that the topology coordinator didn't yet timeout, but will in about 1 second, and we will miss it the fact. And then we will hang here and timeout and fail the test
```python3
        LOGGER.info("Wait until the new node initialization completes or fails.")
        await server_log.wait_for("init - (Startup failed:|Scylla version .* initialization completed)", timeout=120)
```

So here's alternative proposal that should not depend on timing:
let's **concurrently** wait for one of the following events to happen:
- `server_log` prints `"init - (Startup failed:|Scylla version .* initialization completed)"`
- OR `coordinator_log` prints `"The node may hang ..."`

i.e. you spawn concurrent `wait_for` tasks and then wait until first of them finishes.

And then check if it was the coordinator "node may hang", in this case you can shut down the node.

You can also verify that if this happens, the injections was one of the two listed below (if it's a different injection, then that's unexpected behavior and we should fail the test.)

Reply:
OK, got your idea. For now test looks pretty stable, but on following up work on it I will try to improve this part too.